### ISSUE TITLE
DS - SPIKE - Focus Styles

### ISSUE THREAD

**Issue by kellylein** [department-of-veterans-affairs/va-mobile-library#466]
## Description 

As part of this story we want to:   

- DK has been looking into Focus Styles for VADS
- We should look into what is possible with React Native as far as creating custom style(s)
- 

## Acceptance Criteria
  

-  [x] Determine if RN supports custom focus styles 
-  [ ] If we find a way, document what is available to style

[Here is DK's demo. ](https://silver-adventure-6k44jeq.pages.github.io/)
[Here are his notes in Github.](https://github.com/department-of-veterans-affairs/vads-focus-styles)

## Notes & Open Questions

- Dependencies/Roadblocks:
    - Any internal/external dependencies?
    - Test accounts needed?
- Does this require QA?
    
- Dev Notes: 

## Ticket Checklist

- [ ] Acceptance criteria defined
- [ ] Labels added (front-end, back-end, feature)
- [ ] Linked to an Epic

**Comment by TimRoe** [comment]
After conducting an investigation into leveraging accessibility focus styles within the design system, the top level conclusion is as follows:
- No, at present React Native does not support following VADS to have clear focus styles for our components
- However, iOS and Android do appear to have tooling for it so:
   - React Native may support it in the future for iOS/Android (already does for other platforms, e.g. Web, MacOS, Windows)
   - There's potential to revisit this if/when the design system progresses on native integration so it could leverage the native bridge into iOS/Android functions, although it may be complex enough to still not be worthwhile

Additional research details:
- [Pressable](https://reactnative.dev/docs/0.73/pressable) is used for actionable pieces of design system components so far
   - Component contains 4 relevant props: 
	   - onHoverIn and onHoverOut
		   - Did not work with tap actions or screen reader
		   - Did not proceed to testing external keyboard/mouse
		   - Did not work in Web Storybook
	   - onFocus and onBlur
		   - Documentation notes they are for MacOS and Windows
		   - Work in Storybook Web both on manual focus (e.g. clicking button) and loss-of-focus (clicking elsewhere) and tab focus (tabbing on then off)
		   - Do not work in iOS/Android
- [TouchableWithoutFeedback](https://reactnative.dev/docs/next/touchablewithoutfeedback)
	- Component has onFocus/onBlur props like Pressable, but did not have documentation indicating limited usage--testing appeared to behave the same as Pressable
- Came across the [useIsFocused](https://reactnavigation.org/docs/use-is-focused/) hook for react-navigation, but determined it is referring to the screen/app, not a particular element
- Came across a [StackOverflow discussion](https://stackoverflow.com/questions/69591324/react-native-web-expo-how-to-emulate-focus-visible-on-a-pressable) using a different method that relied on the [`react-native-paper` library](https://callstack.github.io/react-native-paper/docs/guides/getting-started/#usage) and included [a testing snack](https://snack.expo.dev/@paulwongx/rnbuttonfocus?platform=web), but this also only appeared to work in web, not iOS/Android
- Came across [another method leveraging hooks and React](https://www.dhiwise.com/post/how-to-leverage-the-useisfocused-hook-in-react-for-improved-user-experience#how-to-determine-if-an-element-is-focused-in-react), but the internals for the hook structure relied on state only available in web, not iOS/Android
- Came across [a general purpose article discussing focus styles including on mobile](https://www.tpgi.com/native-or-custom-a-guide-to-recognizing-focus-styles/), but the only commentary for iOS/Android was updating the OS-level focus settings, nothing on setting within the code
- iOS Native
	- Investigating iOS native support, came across Orange Accessibility pointing to both [Focus an element](https://a11y-guidelines.orange.com/en/mobile/ios/development/#focus-an-element) and [Modify the focus area of VoiceOver](https://a11y-guidelines.orange.com/en/mobile/ios/development/#modify-the-focus-area-of-voiceover) which in turn led to [iOS documentation](https://developer.apple.com/documentation/objectivec/nsobject/uiaccessibilityfocus) for [capturing whether or not an element is focused](https://developer.apple.com/documentation/uikit/uifocussystem) which could then be used to apply custom styling
		- These properties being pulled into React Native code via a hook theoretically could be paired to apply any custom component styling desired
- Android Native
	- Came across a [StackOverflow discussion](https://stackoverflow.com/questions/68852369/how-to-override-focused-button-element-style-android) that pointed to Android build-level customization for global overriding of styles for the app including focus state
		- Following the above, came across [Android Drawable state_focused](https://developer.android.com/guide/topics/resources/drawable-resource#StateList) which could potentially, like with iOS, be made available via hook and theoretically paired to apply any custom component styling desired

So, in conclusion: React Native currently does not provide relevant tooling for focus styling components in iOS/Android. iOS/Android themselves may contain appropriate tooling, but it is not possible currently to access that via the design system and applying at the app-level would likely also be more trouble than it's worth. This research did not investigate a proof-of-concept or prototype proving that the iOS/Android could be leveraged to work, there may be challenges in aligning an element focused at the OS level and what is focused with React Native to apply custom styling.

This conclusion could change in the future as React Native continues to grow and evolve, but for the purpose of this research the conclusion is to revisit in the future. A future point to revisit may be after VADS has fully implemented focus styles which sounds like it may not be a near term goal and buy at least a year to see if there's any movement in React Native.

**Comment by TimRoe** [comment]
Looping in @kellylein @jessicawoodin @brea11y to the comment above on the conclusion of the research.

tl;dr: 
- Nothing in React Native for it
- Some theoretically possible options with iOS/Android, but not something to explore further prior to the design system being capable of native functionality which we deemed appropriate to postpone till some undetermined future date
- Perhaps revisit at the earliest after VADS has finished implementing in web which sounded like it could be a year or more out anyway
